server/server.go

```diff
@@ -247,55 +247,153 @@ func (server *HTTPServer) checkVotePermissions(writer http.ResponseWriter, reque
-		// everything has been handled already
+		handleServerError(writer, err)
 		return
 	}
 
-	userID, successful := server.readUserID(err, request, writer)
+	err = responses.SendResponse(writer, responses.BuildOkResponse())
+	if err != nil {
+		log.Error().Err(err).Msg(responseDataError)
+	}
+}
+
+func (server *HTTPServer) getVoteOnRule(writer http.ResponseWriter, request *http.Request) {
+	clusterID, ruleID, userID, successful := server.readVoteOnRuleParams(writer, request)
 	if !successful {
 		// everything has been handled already
 		return
 	}
 
-	// it's gonna raise an error if cluster does not exist
-	_, _, err = server.Storage.ReadReportForClusterByClusterName(clusterID)
+	userFeedbackOnRule, err := server.Storage.GetUserFeedbackOnRule(clusterID, ruleID, userID)
 	if err != nil {
 		handleServerError(writer, err)
 		return
 	}
 
-	_, err = server.Storage.GetRuleByID(ruleID)
+	err = responses.SendResponse(writer, responses.BuildOkResponseWithData("vote", userFeedbackOnRule.UserVote))
+	if err != nil {
+		log.Error().Err(err).Msg(responseDataError)
+	}
+}
+
+func (server *HTTPServer) createRule(writer http.ResponseWriter, request *http.Request) {
+	ruleID, err := readRuleID(writer, request)
+	if err != nil {
+		// everything has been handled already
+		return
+	}
+
+	var rule types.Rule
+
+	err = json.NewDecoder(request.Body).Decode(&rule)
 	if err != nil {
 		handleServerError(writer, err)
 		return
 	}
 
-	err = server.checkVotePermissions(writer, request, clusterID)
+	rule.Module = ruleID
+
+	err = server.Storage.CreateRule(rule)
+	if err != nil {
+		handleServerError(writer, err)
+		return
+	}
+
+	err = responses.SendResponse(writer, responses.BuildOkResponseWithData(
+		"rule", rule,
+	))
+	if err != nil {
+		log.Error().Err(err).Msg(responseDataError)
+	}
+}
+
+func (server *HTTPServer) createRuleErrorKey(writer http.ResponseWriter, request *http.Request) {
+	ruleID, err := readRuleID(writer, request)
+	if err != nil {
+		// everything has been handled already
+		return
+	}
+
+	errorKey, err := readErrorKey(writer, request)
 	if err != nil {
 		// everything has been handled already
 		return
 	}
 
-	err = server.Storage.VoteOnRule(clusterID, ruleID, userID, userVote)
+	var ruleErrorKey types.RuleErrorKey
+
+	err = json.NewDecoder(request.Body).Decode(&ruleErrorKey)
 	if err != nil {
 		handleServerError(writer, err)
 		return
 	}
 
-	err = responses.SendResponse(writer, responses.BuildOkResponse())
+	ruleErrorKey.RuleModule = ruleID
+	ruleErrorKey.ErrorKey = errorKey
+
+	err = server.Storage.CreateRuleErrorKey(ruleErrorKey)
+	if err != nil {
+		handleServerError(writer, err)
+		return
+	}
+
+	err = responses.SendResponse(writer, responses.BuildOkResponseWithData(
+		"rule_error_key", ruleErrorKey,
+	))
 	if err != nil {
 		log.Error().Err(err).Msg(responseDataError)
 	}
 }
 
+func (server *HTTPServer) readVoteOnRuleParams(writer http.ResponseWriter, request *http.Request) (types.ClusterName, types.RuleID, types.UserID, bool) {
```

Comment:
I think we should extract functions like this to separate file, and in `server.go` leave only endpoints

Reply:
orgId don't have to be handled. It's per-user vote

Reply:
@tisnik in that case how do we check permissions? What if user orgId and cluster orgId don't match?

Reply:
Found it

Reply:
done